File: rc4.c
#include <stdio.h>
#include <string.h>

// Function for the Key Scheduling Algorithm (KSA)
void KSA(unsigned char S[], unsigned char key[], int key_len, int N) {
    int i, j;
    j = 0;
    for (i = 0; i < N; i++) {
        j = (j + S[i] + key[i % key_len]) % N;
        // Swap S[i] and S[j]
        unsigned char temp = S[i];
        S[i] = S[j];
        S[j] = temp;
    }
}

// Function for the Pseudo-Random Generation Algorithm (PGRA)
void PGRA(unsigned char S[], unsigned char data[], int data_len, int N) {
    int i, j;
    i = j = 0;
    for (int k = 0; k < data_len; k++) {
        i = (i + 1) % N;
        j = (j + S[i]) % N;
        // Swap S[i] and S[j]
        unsigned char temp = S[i];
        S[i] = S[j];
        S[j] = temp;
        int t = (S[i] + S[j]) % N;
        data[k] ^= S[t]; // XOR operation with the state vector
    }
}

int main() {
    unsigned char S[256]; // State vector for RC4
    unsigned char key[] = "SecretKey"; // Encryption key
    unsigned char input_text[] = "Hello, RC4!"; // Input text
    unsigned char ciphertext[256]; // Encrypted data (ciphertext)
    unsigned char decrypted_data[256]; // Decrypted data

    int N = 256; // Number of elements in the state vector

    int key_len = strlen(key);
    int text_len = strlen(input_text);

    printf("Input text: %s\n", input_text);
    printf("Key: %s\n\n", key);

    // Initialize S
    for (int i = 0; i < N; i++) {
        S[i] = i;
    }

    // Perform the Key Scheduling Algorithm (KSA) for both encryption and decryption
    KSA(S, key, key_len, N);

    // Perform encryption using PGRA
    PGRA(S, input_text, text_len, N);

    printf("Encryption complete.\n");

    // Display ciphertext
    printf("Ciphertext: ");
    for (int i = 0; i < text_len; i++) {
        printf("%02X ", input_text[i]);
    }
    printf("\n\n");

    // Reset S
    for (int i = 0; i < N; i++) {
        S[i] = i;
    }

    // Perform the Key Scheduling Algorithm (KSA) for decryption
    KSA(S, key, key_len, N);

    // Perform decryption using PGRA
    PGRA(S, input_text, text_len, N);

    // Display decrypted data
    printf("Decrypted data: %s\n", input_text);

    return 0;
}

File: rc4.rs
use std::io;

// Function for the Key Scheduling Algorithm (KSA)
fn ksa(s: &mut [u8; 256], key: &[u8], key_len: usize, n: usize) {
    let mut j = 0;
    for i in 0..n {
        j = (j + s[i] as usize + key[i % key_len] as usize) % n;
        // Swap s[i] and s[j]
        s.swap(i, j);
    }
}

// Function for the Pseudo-Random Generation Algorithm (PGRA)
fn pgra(s: &mut [u8; 256], data: &mut [u8], data_len: usize, n: usize) {
    let mut i = 0;
    let mut j = 0;
    for k in 0..data_len {
        i = (i + 1) % n;
        j = (j + s[i] as usize) % n;
        // Swap s[i] and s[j]
        s.swap(i, j);
        let t = (s[i] as usize + s[j] as usize) % n;
        data[k] ^= s[t]; // XOR operation with the state vector
    }
}

fn main() {
    let mut s: [u8; 256] = [0; 256]; // State vector for RC4
    let key: [u8; 9] = *b"SecretKey"; // Encryption key
    let mut input_text: Vec<u8> = "Hello, RC4!".as_bytes().to_vec(); // Input text
    let mut ciphertext: Vec<u8> = vec![0; input_text.len()]; // Encrypted data (ciphertext)
    let mut decrypted_data: Vec<u8> = vec![0; input_text.len()]; // Decrypted data

    let n = 256; // Number of elements in the state vector

    let key_len = key.len();
    let text_len = input_text.len();

    println!("Input text: {}", String::from_utf8_lossy(&input_text));
    println!("Key: {}\n", String::from_utf8_lossy(&key));

    // Initialize S
    for i in 0..n {
        s[i] = i as u8;
    }

    // Perform the Key Scheduling Algorithm (KSA) for both encryption and decryption
    ksa(&mut s, &key, key_len, n);

    // Perform encryption using PGRA
    pgra(&mut s, &mut input_text, text_len, n);

    println!("Encryption complete.");

    // Display ciphertext
    print!("Ciphertext: ");
    for byte in &input_text {
        print!("{:02X} ", byte);
    }
    println!("\n");

    // Reset S
    for i in 0..n {
        s[i] = i as u8;
    }

    // Perform the Key Scheduling Algorithm (KSA) for decryption
    ksa(&mut s, &key, key_len, n);

    // Perform decryption using PGRA
    pgra(&mut s, &mut input_text, text_len, n);

    // Display decrypted data
    println!("Decrypted data: {}", String::from_utf8_lossy(&input_text));
}
